src/core/channels/presence_channel.ts

```diff
@@ -41,12 +42,28 @@ export default class PresenceChannel extends PrivateChannel {
     });
   }
 
-  /** Handles presence and subscription events. For internal use only.
+  /** Handles presence and subscription messages. For internal use only.
    *
-   * @param {String} event
-   * @param {*} data
+   * @param {Message} message
    */
-  handleEvent(event : string, data : any) {
+  handleMessage(message: Message) {
+    var event = message.event;
+    if (event.indexOf("pusher_internal:") === 0) {
+      this.handleInternalMessage(message)
+    } else {
+      var data = message.data;
+      if (message.user_id) {
+        var metadata = {};
+        metadata['user_id'] = message.user_id;
+        this.emit(event, data, metadata);
+      } else {
+        this.emit(event, data);
+      }
+    }
+  }
+  handleInternalMessage(message: Message) {
+    var event = message.event;
+    var data = message.data;
     switch (event) {
       case "pusher_internal:subscription_succeeded":
```

Comment:
This code is a duplicate of `Channel.handleMessage`, except for the call to `members.onSubscription`. Can we reduce the duplication?

Reply:
Tried to reduce the duplication with https://github.com/pusher/pusher-js/pull/323/commits/86936530cbb2d582c16e51c5b854ecedd0d7ca7c